test/e2eexternal/dataplane/connectivity.go

```diff
@@ -0,0 +1,335 @@
+		// source IP seems to become the node IP of external network CIDRs, like 172.16.0.3
+		// not the dockerIP (external app's IP). This needs to be investigated further, before merge.
+		if framework.TestContext.ClusterIDs[idx] != "cluster1" {
+			Expect(dockerLog).To(ContainSubstring(podIP))
+		}
+		// temporary debug log
+		framework.Logf("dockerLog: %v", dockerLog)
+	}
+}
+
+func makePodSpec(nameBase string, label map[string]string, containerName, image string, command []string) *v1.Pod {
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: nameBase,
+			Labels:       label,
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:    containerName,
+					Image:   image,
+					Command: command,
+				},
+			},
+		},
+	}
+}
+
+func createPod(cs *kubeclientset.Clientset, namespace string, podSpec *v1.Pod) (*v1.Pod, error) {
+	pod, err := cs.CoreV1().Pods(namespace).Create(podSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	return pod, nil
+}
+
+func waitForPodReady(cs *kubeclientset.Clientset, namespace, name string) error {
+	return wait.PollImmediate(
+		5*time.Second,  /* interval */
+		60*time.Second, /* timeout */
+		func() (bool, error) {
+			pod, err := cs.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
+			if err != nil {
+				return false, nil
+			}
+			switch pod.Status.Phase {
+			case v1.PodRunning:
+				for _, status := range pod.Status.ContainerStatuses {
+					if !status.Ready {
+						return false, nil
+					}
+				}
+				return true, nil
+			case v1.PodFailed, v1.PodSucceeded:
+				return false, fmt.Errorf("pod won't become running (current phase: %v)", pod.Status.Phase)
+			}
+
+			return false, nil
+		})
+}
+
+func makeServiceSpec(svcName string, label map[string]string, port, targetPort int) *v1.Service {
+	return &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: svcName,
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{{
+				Port:       int32(port),
+				TargetPort: intstr.FromInt(targetPort),
+				Protocol:   v1.ProtocolTCP,
+			}},
+			Selector: label,
+		},
+	}
+}
+
+func createService(cs *kubeclientset.Clientset, namespace string, svcSpec *v1.Service) (*v1.Service, error) {
+	svc, err := cs.CoreV1().Services(namespace).Create(svcSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	return svc, nil
+}
+
+func sendHTTPRequestFromDocker(name, ip string, port int) error {
+	command := []string{"curl", "-m", "10", fmt.Sprintf("%s:%d", ip, port)}
+	return runCommandInDocker(name, command...)
+}
+
+func runCommandInDocker(name string, command ...string) error {
+	var stdout, stderr bytes.Buffer
+
+	cmdargs := []string{"exec", "-i", name}
+	cmdargs = append(cmdargs, command...)
+	cmd := exec.Command("docker", cmdargs...)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to run command docker %v: %v, stdout: %v, stderr: %v",
+			cmdargs, err, stdout.String(), stderr.String())
+	}
+
+	return nil
+}
+
```

Comment:
Let's make a docker.go to separate the docker handling functions from the test logic